src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,267 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {GetRowsOptions, PrefixRange} from '../table';
+import {ChunkTransformer} from '../chunktransformer';
+import * as protos from '../../protos/protos';
+import {TableUtils} from './table';
+import {google} from '../../protos/protos';
+import {CallOptions, GoogleError, RetryOptions, ServiceError} from 'google-gax';
+import {Mutation} from '../mutation';
+import {BoundData, Filter} from '../filter';
+import {RequestType} from 'google-gax/build/src/apitypes';
+import {
+  DEFAULT_BACKOFF_SETTINGS,
+  RETRYABLE_STATUS_CODES,
+} from './retry-options';
+import * as is from 'is';
+
+// This interface contains the information that will be used in a request.
+interface TableStrategyInfo {
+  tableName: string;
+  appProfileId?: string;
+}
+
+// Gets the row keys for a readrows request.
+function getRowKeys(
+  rowKeys: string[],
+  lastRowKey: string | number | true | Uint8Array
+) {
+  // Remove rowKeys already read.
+  return rowKeys.filter(rowKey =>
+    TableUtils.greaterThan(rowKey, lastRowKey as string)
+  );
+}
+
+// Modifies ranges in place based on the lastRowKey to prepare
+// a readrows request.
+function spliceRanges(
+  ranges: PrefixRange[],
+  lastRowKey: string | number | true | Uint8Array
+): void {
+  // Readjust and/or remove ranges based on previous valid row reads.
+  // Iterate backward since items may need to be removed.
```

Comment:
Request for clarification - why do you have to iterate backwards? (this might be obvious to a CBT person but it's not to me)

Reply:
This code was copied from [here](https://github.com/googleapis/nodejs-bigtable/blob/b2b2fdae61865a36efc8f1f077276053f41d992f/src/table.ts#L848). I think it is because the splice operation is faster if removal is done to the end of the array before the beginning.